src/datasource.test.ts

```diff
@@ -1,32 +1,145 @@
-  });
-  expect(response).toBeTruthy();
+  const response = await ds.query(query);
+
+  expect(response.data).toHaveLength(0);
 });
 
+it('should return data ignoring invalid queries', async () => {
+  const query = buildQuery([
+    { refId: 'A', tableId: '_' }, // invalid
+    { refId: 'B', tableId: '1', columns: [{ name: 'float', dataType: 'FLOAT32', columnType: 'NORMAL' }] },
+  ]);
+
+  await ds.query(query);
+
+  expect(fetchMock).toBeCalledTimes(1);
+  expect(fetchMock).toBeCalledWith(expect.objectContaining({ url: '_/v1/tables/1/query-data' }));
+});
+
+it('should return data for multiple targets', async () => {
+  const query = buildQuery([
+    { refId: 'A', tableId: '1', columns: [{ name: 'int', dataType: 'INT32', columnType: 'NORMAL' }] },
+    { refId: 'B', tableId: '2', columns: [{ name: 'float', dataType: 'FLOAT32', columnType: 'NORMAL' }] },
+  ]);
+
+  const response = await ds.query(query);
+
+  expect(fetchMock).toBeCalledTimes(2);
+  expect(response.data).toHaveLength(2);
+});
+
+it('should convert columns to Grafana fields', async () => {
+  const query = buildQuery([
+    {
+      refId: 'A',
+      tableId: '_',
+      columns: [
+        { name: 'int', dataType: 'INT32', columnType: 'INDEX' },
+        { name: 'float', dataType: 'FLOAT32', columnType: 'NORMAL' },
+        { name: 'string', dataType: 'STRING', columnType: 'NORMAL' },
+        { name: 'time', dataType: 'TIMESTAMP', columnType: 'NORMAL' },
+        { name: 'bool', dataType: 'BOOL', columnType: 'NORMAL' },
+      ],
+    },
+  ]);
+
+  const response = await ds.query(query);
+
+  const fields = response.data[0].fields as Field[];
+  const actual = fields.map(({ name, type, values }) => ({ name, type, values: values.toArray() }));
+  expect(actual).toEqual([
+    { name: 'int', type: FieldType.number, values: [1, 2] },
+    { name: 'float', type: FieldType.number, values: [1.1, 2.2] },
+    { name: 'string', type: FieldType.string, values: ['first', 'second'] },
+    { name: 'time', type: FieldType.time, values: [1663135260000, 1663135320000] },
+    { name: 'bool', type: FieldType.boolean, values: [true, false] },
+  ]);
+});
+
+it('should automatically apply time filters when index column is a timestamp', async () => {
+  const query = buildQuery([
+    { refId: 'A', tableId: '_', columns: [{ name: 'time', dataType: 'TIMESTAMP', columnType: 'INDEX' }] },
+  ]);
+  const from = dateTime('2022-09-14T00:00:00Z');
+  const to = dateTime('2022-09-16T00:00:00Z');
+  query.range = { from, to, raw: { from, to } };
+
+  await ds.query(query);
+
+  expect(fetchMock).toBeCalledWith(
+    expect.objectContaining({
+      data: expect.objectContaining({
+        filters: [
+          { column: 'time', operation: 'GREATER_THAN_EQUALS', value: '2022-09-14T00:00:00.000Z' },
+          { column: 'time', operation: 'LESS_THAN_EQUALS', value: '2022-09-16T00:00:00.000Z' },
+        ],
+      }),
```

Comment:
```suggestion
  const query = buildQuery([
    { refId: 'A', tableId: '_', columns: [{ name: 'time', dataType: 'TIMESTAMP', columnType: 'INDEX' }] },
  ]);
  const from = '2022-09-14T00:00:00Z';
  const to = '2022-09-16T00:00:00Z';
  query.range = { dateTime(from), dateTime(to), raw: { from, to } };

  await ds.query(query);

  expect(fetchMock).toBeCalledWith(
    expect.objectContaining({
      data: expect.objectContaining({
        filters: [
          { column: 'time', operation: 'GREATER_THAN_EQUALS', value: from },
          { column: 'time', operation: 'LESS_THAN_EQUALS', value: to },
        ],
      }),
```

I don't know if the above is valid typescript, but you get the idea. It'd be nice to have a constant for `2022-09-14T00:00:00Z`, is all I'm getting at.

Reply:
I took a different approach, but still removed the duplication.